Vignette: Data Preparation and App Deployment

nandriychuk opened the pull request:
## Overview
This branch adds the Data Preparation and App Deployment Vignette to the package 

## Connected Issues
- Closes #195 
- Closes #414

jharmon-gilead committed:
Merge pull request #481 from Gilead-BioStats/dev

{gsm.app} v2.5.1 RC

Commit:
data-prep vignette first draft

Commit:
update req data scructures

Commit:
add workflow, fetchData and runApp steps

Commit:
add deployment section

nandriychuk committed:
update chunks

jharmon-gilead committed:
Merge branch 'dev' into fix-195

jonthegeek commented on _pkgdown.yml:
```diff
@@ -1,3 +1,17 @@
+
+navbar:
+  structure:
+    left: [intro, reference, articles, tutorials, news]
```
```suggestion
    left: [intro, reference, articles, news]
```
We don't have any tutorial content (hopefully this is "yet", but let's not put in an empty section in this PR; we'll add it when it's useful). Overall it looks like this is restating the deffault layout, so I'm pretty sure we just just delete this block entirely, actually.

jonthegeek commented on _pkgdown.yml:
```diff
@@ -1,3 +1,17 @@
+    left: [intro, reference, articles, tutorials, news]
+    right: [search, github]
+
+articles:
+- title: Getting Started
+  navbar: ~
+  contents:
+  - data-preparation
+- title: Advanced Topics  
+  contents:
+  - plugins
```
```suggestion
```
Until we have a very good reason to do so, don't specify what goes where. Once we start doing this, we have the extra step of adding a thing here when we add a vignette. If we leave this off, pkgdown will simply include every vignette, without us having to do an extra step.

jonthegeek commented on vignettes/data-preparation.Rmd:
````diff
@@ -0,0 +1,399 @@
+)
+```
+
+```{r setup, include = FALSE}
````
````suggestion
```{r setup}
````
We want to show this to users so they know which packages are needed. With include = FALSE and the `eval = FALSE` option, this chunk doesn't do anything.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+
+The image below illustrates the overarching context in which the reporting workflow runs, taking inputs from both the output of the analytics workflow, as well as raw study-, site-, and country-level data in the Raw/Raw+ format.
+
+![](data_model_detailed.png){width="100%"}
```
This image doesn't exist.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+
+![](data_model_detailed.png){width="100%"}
+
+# Required Data Structures
```
I... feel like all of this has too much detail. We should point to the other gsm documentation about preparing data. The unique requirements that we add are `fnFetchData()` and (optionally) `fnCountData()`. Everything from here through line 304 should be condensed to a brief description and a link to probably {gsm.core} documentation.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+    strSubjectID = strSubjectID,
+    dSnapshotDate = dSnapshotDate
+  )
+}
```
This is the important part. Anonymize an actual working example from Spencer of how we really do it, and show that. The sample function is very very sample-y and is unlikely to match real use.

Try to minimize comments in the code blocks. We're in an Rmd, so we can and should write non-code outside of the blocks.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+
+Once you have prepared all the required data structures and the `fnFetchData` function, you can deploy the {gsm.app}.
+
+For development and testing, run the app locally: run_gsm_app() creates a Shiny app to explore a set of clinical trial data. The app facilitates exploration of the data by allowing the user to click to diver deeper into aspects of the data.
```
```suggestion
For development and testing, run the app locally: run_gsm_app() creates a Shiny app to explore a set of clinical trial data. The app facilitates exploration of the data by allowing the user to click to dive deeper into aspects of the data.
```
Typo. But I'm not sure this statement fits at this point in the discussion.

I'd refocus all of this specifically on preparing the data.frames (via reference to {gsm.core}), writing `fnFetchData()` and `fnCountData()`, and then launching the app locally. We can briefly mention production deployment, but that's another step above the focus here and will be very specific to the user's context.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+
+## Common Issues
+
+1.  **Data Validation Errors**: Ensure all required columns are present and correctly named
```
This likely won't happen, but it can be used as another plug to use the standard workflows.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+## Common Issues
+
+1.  **Data Validation Errors**: Ensure all required columns are present and correctly named
+2.  **Missing MetricID Links**: Verify MetricID values are consistent across all data frames
```
This shouldn't ever happen if they're using gsm workflows. I wouldn't consider this a common issue.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+
+1.  **Data Validation Errors**: Ensure all required columns are present and correctly named
+2.  **Missing MetricID Links**: Verify MetricID values are consistent across all data frames
+3.  **Date Format Issues**: Ensure all date columns are properly formatted as Date objects
```
```suggestion
3.  **Date Format Issues**: Ensure all date columns are properly formatted as Date objects
```
I don't think this is ever really an issue in the core app. gsm.ae cares about dates but gsm.app doesn't in any way I can think of.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+1.  **Data Validation Errors**: Ensure all required columns are present and correctly named
+2.  **Missing MetricID Links**: Verify MetricID values are consistent across all data frames
+3.  **Date Format Issues**: Ensure all date columns are properly formatted as Date objects
+4.  **Memory Issues**: For large datasets, consider data filtering and optimization
```
Make this sound more specific to the type of data we're looking at ("For extremely large clinical trials" or something along those lines), and offer more specific tips. I'm actually not sure what's valid as far as subdividing the core data. Realistically this is probably only really a potential issue for domain data. We pass errors on from `fnFetchData()` to users, so (here or in the `fnFetchData()` section), for example, I'd recommend having fnFetchData() throw an error suggesting that the user choose a site and/or participant if those values come to fnFetchData as NULL and the data is too big to load.

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+2.  **Missing MetricID Links**: Verify MetricID values are consistent across all data frames
+3.  **Date Format Issues**: Ensure all date columns are properly formatted as Date objects
+4.  **Memory Issues**: For large datasets, consider data filtering and optimization
+5.  **Fetch Function Errors**: Test your `fnFetchData` function with various parameter combinations
```
Mention that errors are passed on to the user through the UI. I'd put this specifically into the fnFetchData() section and leave it off as a common issue. I'm... actually not certain that any of these ARE common issues. At most we have "Things we've noticed in our pilots" so far...

jonthegeek commented on vignettes/data-preparation.Rmd:
```diff
@@ -0,0 +1,399 @@
+5.  **Fetch Function Errors**: Test your `fnFetchData` function with various parameter combinations
+
+
+# Summary
+
+This vignette has demonstrated a comprehensive workflow for preparing clinical trial data for use with {gsm.app} using the {gsm.core} and {gsm.reporting} packages. The approach provides:
+
+## Key Benefits
+
+1.  **Validated Statistical Methods**: Uses peer-reviewed KRI calculation methods from {gsm.core}
+2.  **Standardized Data Model**: Ensures compatibility across the {gsm} ecosystem
+3.  **Professional Reporting**: Produces publication-ready visualizations and tables
+4.  **Automated Workflows**: Reduces manual intervention and potential errors
+5.  **Quality Assurance**: Built-in validation and data quality checks
+6.  **Scalable Architecture**: Suitable for both development and production environments
+
+## Next Steps
+
+-   **Customize Workflows**: Modify YAML configuration files to match your specific study requirements
+-   **Validate Results**: Use the validation functions provided in {gsm.core} to verify data quality
+-   **Deploy Applications**: Choose the deployment method that best fits your infrastructure needs
+-   **Monitor Performance**: Implement the performance optimization techniques for larger datasets
+
+## Additional Resources
+
+-   [gsm.core Documentation](https://gilead-biostats.github.io/gsm.core/)
+-   [gsm.reporting Documentation](https://gilead-biostats.github.io/gsm.reporting/)
+-   [gsm.app Repository](https://github.com/Gilead-BioStats/gsm.app)
+
+For questions or issues, please refer to the package documentation or submit issues to the appropriate GitHub repositories.
```
This feels generic to the gsm family of packages. I don't think it belongs in this vignette.

jonthegeek reviewed (changes requested):
Overall, I'd like to make this much more focused. A single large vignette isn't useful; it makes it hard to find the specific thing the user is struggling with. The more I think about it, the clearer it becomes that this vignette should focus only on `fnFetchData()` (and maybe `fnCountData()`). If we feel like the other pieces still need more explanation than we can give in the help for `run_gsm_app()`, we can add separate vignettes.

What goes into `fnFetchData()` is probably fairly non-obvious from the existing help and code, so you'll need to discuss it with Spencer, and make sure you understand how it should work (and why we need the user to provide it).

nandriychuk commented:
> Overall, I'd like to make this much more focused. A single large vignette isn't useful; it makes it hard to find the specific thing the user is struggling with. The more I think about it, the clearer it becomes that this vignette should focus only on `fnFetchData()` (and maybe `fnCountData()`). If we feel like the other pieces still need more explanation than we can give in the help for `run_gsm_app()`, we can add separate vignettes.
> 
> What goes into `fnFetchData()` is probably fairly non-obvious from the existing help and code, so you'll need to discuss it with Spencer, and make sure you understand how it should work (and why we need the user to provide it).

ok, I'll set up a meeting with @samussiah to discuss the `fnFetchData()` and `fnCountData()` functions. You are correct, it wasn't straightforward to me the way they work. 
As for the existing vignette in the PR, we can break it apart into smaller chunks or abandon completely if needed. One thing that was helpful for me as a programmer who never used gsm.app before is to figure out how all pieces of the gsm unfrastructure work together since it wasn't obvious. It took me some time to actually figure out how mutliple packages work and the sequence they should be used in order to prepare the data to be used in the app. Hoever, if we think that the average gsm.app user would already have these knowledge coming into configuring the app, the vignette of this type won't be helpful.

nandriychuk commented:
@jonthegeek, thank you for your review! I read all comments and going to work on these!

jonthegeek commented:
> One thing that was helpful for me as a programmer who never used gsm.app before is to figure out how all pieces of the gsm unfrastructure work together since it wasn't obvious. It took me some time to actually figure out how mutliple packages work and the sequence they should be used in order to prepare the data to be used in the app. Hoever, if we think that the average gsm.app user would already have these knowledge coming into configuring the app, the vignette of this type won't be helpful.

I just don't think it's {gsm.app}-specific. Turning raw data into those tibbles is a broader gsm.core activity, so we should probably (slightly) expand the help for those fields in `run_gsm_app()` (as defined in `R/aaa-shared.R`) to point users to relevant topics elsewhere in gsm, but gsm.app doesn't "own" how they work, so it shouldn't go deep into details. If they are changed elsewhere in gsm, we will update gsm.app to work with those changes, but we won't expect the changers to come find how we describe them in gsm.app and update the help here as well.

nandriychuk commented:
> > One thing that was helpful for me as a programmer who never used gsm.app before is to figure out how all pieces of the gsm unfrastructure work together since it wasn't obvious. It took me some time to actually figure out how mutliple packages work and the sequence they should be used in order to prepare the data to be used in the app. Hoever, if we think that the average gsm.app user would already have these knowledge coming into configuring the app, the vignette of this type won't be helpful.
> 
> I just don't think it's {gsm.app}-specific. Turning raw data into those tibbles is a broader gsm.core activity, so we should probably (slightly) expand the help for those fields in `run_gsm_app()` (as defined in `R/aaa-shared.R`) to point users to relevant topics elsewhere in gsm, but gsm.app doesn't "own" how they work, so it shouldn't go deep into details. If they are changed elsewhere in gsm, we will update gsm.app to work with those changes, but we won't expect the changers to come find how we describe them in gsm.app and update the help here as well.

That makes a lot of sense, I can definitely see your point.

nandriychuk committed:
address feedback

nandriychuk committed:
update vignette with fetchData and countData

nandriychuk reviewed (commented):
Add sentence about using gsm.mapping as a spec and that app is flexible to accept data

nandriychuk committed:
add sentence about gsm.mapping

samussiah committed:
add dedicated data ingestion vignette

samussiah committed:
add "why"

samussiah commented:
@nandriychuk - the examples look good.  I added a dedicated data ingestion vignette that goes into more detail and overlaps with the data preparation vignette a bit.  If you'll review the new vignette and clean up any duplication I think this one's good to go.

jonthegeek committed:
Tweak vignettes.

I moved the `fnCountData()` section over to the `fnFetchData()`
vignette, and, importantly, corrected it. By default, it's automatically
a wrapper around `fnFetchData()` A significant optimization step is
available by providing a unique `fnCountData()` function that has the
database do the counting, rather than returning all results and then
counting. We definitely should implement this for our own
deployments!

jonthegeek committed:
Delete unused image.

jonthegeek reviewed (commented):
These look great, but I want to wait for #539 to be merged before I merge these, so I can reconcile any merge conflicts (and generally improve the pkgdown setup). @samussiah please see my changes to the bit about `fnCountData()`, cuz I think we can likely get a significant speed increase!

jonthegeek committed:
Merged origin/dev into fix-195

jonthegeek committed:
Better articles nav.

jonthegeek committed:
Break off Qualification section more clearly, even though it only contains 1 article.

jonthegeek committed:
Don't show BR article in navbar.

jonthegeek reviewed (approved):
That should do it (once checks pass)!

samussiah commented:
Thanks for the review!
